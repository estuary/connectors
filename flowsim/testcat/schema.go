package testcat

import (
	"reflect"
	"strings"
	"time"
)

// CustomSchema is an interface that allows a struct to build it's own schema
// rather than use the default schema builder.
type CustomBuildSchema interface {
	BuildSchema() (TestSchema, []string, error)
}

// BuildSchema parses a struct and builds the flow.yaml schema representation along with it's keys.
// It can handle just about any struct. To determine the name of the fields in the schema, you can use the
// flowsim tag to set the name. You can also use the 'key' tag option to set the struct field as a key and
// 'required' to set it as a required field. (keys are automatically required) See testdata.Basic for
// and example struct.
func BuildSchema(in interface{}) (TestSchema, []string, error) {

	// If it supports the CustomBuildSchema interface, use it.
	if cbs, ok := in.(CustomBuildSchema); ok {
		return cbs.BuildSchema()
	}

	var schema TestSchema
	var keys []string
	var err error

	// Handle any special built-in types such as time.Time.
	switch in.(type) {
	case time.Time:
		schema.Type = []string{"string"}
		schema.Format = "date-time"
		return schema, keys, err
	}

	value := reflect.ValueOf(in)
	valueKind := value.Kind()

	// Dereference if a pointer.
	if (valueKind == reflect.Ptr && !value.IsNil()) || valueKind == reflect.Interface {
		value = value.Elem()
		valueKind = value.Kind()
	}

	switch valueKind {
	// Support maps, but it won't have any keys.
	case reflect.Map:
		schema.Type = []string{"object"}
		schema.Properties = make(map[string]TestSchema)
		for _, mapKey := range value.MapKeys() {
			if schema.Properties[mapKey.String()], _, err = BuildSchema(value.MapIndex(mapKey).Interface()); err != nil {
				return schema, keys, err
			}
		}
	// For structs, parse all the fields.
	case reflect.Struct:
		schema.Type = []string{"object"}
		schema.Properties = make(map[string]TestSchema)
		for i := 0; i < value.NumField(); i++ {
			structKey := value.Type().Field(i)

			// Non exported fields should be skipped.
			if structKey.PkgPath != "" {
				continue
			}

			// Parse the flowsim tags to find required fields and keys.
			fieldName := structKey.Tag.Get("flowsim")
			tagOptions := []string{}
			if fieldName == "" {
				fieldName = structKey.Name
			} else {
				// Get the field name and tagOptions.
				if idx := strings.Index(fieldName, ","); idx != -1 {
					tagOptions = strings.Split(fieldName[idx+1:], ",")
					fieldName = fieldName[:idx]
				}
				// Skipped field.
				if fieldName == "-" {
					continue
				}
			}
			// If it's a key, mark it as key and as required.
			if optionsHas(tagOptions, "key") {
				keys = append(keys, "/"+fieldName)
				schema.Required = append(schema.Required, fieldName)
			} else if optionsHas(tagOptions, "required") {
				// Mark as required
				schema.Required = append(schema.Required, fieldName)
			}
			// Parse the field properties (as well as child objects).
			var fieldKeys []string
			if schema.Properties[fieldName], fieldKeys, err = BuildSchema(value.Field(i).Interface()); err != nil {
				return schema, keys, err
			}
			// Any child keys need to be appended to the keys slice.
			for _, key := range fieldKeys {
				keys = append(keys, "/"+fieldName+key)
			}
		}
	default:
		// All other types, use reflect.Kind.
		schema.Type = SchemaType(valueKind)
	}

	return schema, keys, nil

}

// optionHas is a helper to parse struct tags and see if it has an option.
func optionsHas(options []string, value string) bool {
	for _, s := range options {
		if s == value {
			return true
		}
	}
	return false
}

// SchemaType returns the json type for a field.
func SchemaType(t reflect.Kind) []string {

	switch t {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return []string{"integer"}
	case reflect.Array, reflect.Slice:
		return []string{"array"}
	case reflect.Float32, reflect.Float64:
		return []string{"number"}
	case reflect.Bool:
		return []string{"boolean"}
	case reflect.String:
		return []string{"string"}
	}
	return []string{"unknown"}

}

// BuildCollection is a helper to automatically build a schema, parse keys and return a Collection.
func BuildCollection(in interface{}) (TestCollection, error) {
	var c TestCollection
	var err error
	if c.Schema, c.Keys, err = BuildSchema(in); err != nil {
		return c, err
	}
	return c, err
}
