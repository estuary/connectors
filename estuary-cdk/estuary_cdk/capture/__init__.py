from dataclasses import dataclass
from pydantic import Field
from typing import Generic, Awaitable, Any, BinaryIO, Callable
from logging import Logger
import abc
import asyncio
import shutil
import sys
import tempfile
import traceback
import xxhash

from . import request, response
from .. import BaseConnector, Stopped
from ..flow import (
    ConnectorSpec,
    ConnectorState,
    ConnectorStateUpdate,
    EndpointConfig,
    ResourceConfig,
)
from ..pydantic_polyfill import GenericModel


class Request(GenericModel, Generic[EndpointConfig, ResourceConfig, ConnectorState]):
    spec: request.Spec | None = None
    discover: request.Discover[EndpointConfig] | None = None
    validate_: request.Validate[EndpointConfig, ResourceConfig] | None = Field(
        default=None, alias="validate"
    )
    apply: request.Apply[EndpointConfig, ResourceConfig] | None = None
    open: request.Open[EndpointConfig, ResourceConfig, ConnectorState] | None = None
    acknowledge: request.Acknowledge | None = None


class Response(GenericModel, Generic[EndpointConfig, ResourceConfig, ConnectorState]):
    spec: ConnectorSpec | None = None
    discovered: response.Discovered[ResourceConfig] | None = None
    validated: response.Validated | None = None
    applied: response.Applied | None = None
    opened: response.Opened | None = None
    captured: response.Captured | None = None
    checkpoint: response.Checkpoint[ConnectorState] | None = None


@dataclass
class Task:
    """
    Task bundles a capture coroutine and its associated context.

    It facilitates the task in queuing any number of captured documents and
    emitting them upon a checkpoint. Each instance of Task manages an independent
    internal buffer (backed by memory or disk) which is only written to the
    connector's output upon a call to checkpoint(). This allows concurrent
    Tasks to capture consistent checkpoints without trampling one another.

    Task also facilitates logging and graceful stop of a capture coroutine.
    """

    log: Logger
    """Attached Logger of this Task instance, to use for scoped logging."""

    @dataclass
    class Stopping:
        """
        Stopping coordinates the graceful exit of capture Tasks.

        Its `event` is set when Tasks should gracefully stop.
        The Task's coroutine should monitor this event and exit when it's set AND
        it has no more immediate work to do (for example, no further documents are
        currently ready to be captured).
        """

        event: asyncio.Event
        first_error: Exception | None = None
        first_error_task: str | None = None

    stopping: Stopping

    _buffer: tempfile.SpooledTemporaryFile
    _hasher: xxhash.xxh3_128
    _name: str
    _output: BinaryIO
    _tg: asyncio.TaskGroup

    MAX_BUFFER_MEM: int = 1_000_000
    """Maximum amount of memory to use for captured documents between checkpoints,
    before spilling to disk."""

    def __init__(
        self,
        log: Logger,
        name: str,
        output: BinaryIO,
        stopping: Stopping,
        tg: asyncio.TaskGroup,
    ):
        self._buffer = tempfile.SpooledTemporaryFile(max_size=self.MAX_BUFFER_MEM)
        self._hasher = xxhash.xxh3_128()
        self._name = name
        self._output = output
        self._tg = tg
        self.log = log
        self.stopping = stopping

    def captured(self, binding: int, document: Any):
        """Enqueue the document to be captured under the given binding.
        Documents are not actually captured until checkpoint() is called.
        Or, reset() will discard any queued documents."""

        b = Response(
            captured=response.Captured(binding=binding, doc=document)
        ).model_dump_json(by_alias=True, exclude_unset=True)

        self._buffer.write(b.encode())
        self._buffer.write(b"\n")
        self._hasher.update(b)

    def pending_digest(self) -> str:
        """pending_digest returns the digest of captured() documents
        since the last checkpoint() or reset()"""

        return self._hasher.digest().hex()

    def checkpoint(self, state: ConnectorState, merge_patch: bool = True):
        """Emit previously-queued, captured documents follows by a checkpoint"""

        self._emit(
            Response[Any, Any, ConnectorState](
                checkpoint=response.Checkpoint(
                    state=ConnectorStateUpdate(updated=state, mergePatch=merge_patch)
                )
            )
        )

    def reset(self):
        """Discard any captured documents, resetting to an empty state."""
        self._buffer.truncate(0)
        self._buffer.seek(0)
        self._hasher.reset()

    def spawn_child(
        self, name_suffix: str, child: Callable[["Task"], Awaitable[None]]
    ) -> asyncio.Task:
        """
        Spawn a child Task of this Task, using the given name suffix and coroutine.
        The child coroutine will be invoked with a child Task and be polled concurrently.
        """

        child_name = f"{self._name}.{name_suffix}"
        child_log = self.log.getChild(name_suffix)

        async def run_task(parent: Task):
            async with asyncio.TaskGroup() as child_tg:
                try:
                    t = Task(
                        child_log,
                        child_name,
                        parent._output,
                        parent.stopping,
                        child_tg,
                    )
                    await child(t)
                except Exception as exc:
                    child_log.error("".join(traceback.format_exception(exc)))

                    if parent.stopping.first_error is None:
                        parent.stopping.first_error = exc
                        parent.stopping.first_error_task = child_name

                    parent.stopping.event.set()

        task = self._tg.create_task(run_task(self))
        task.set_name(child_name)
        return task

    def _emit(self, response: Response[EndpointConfig, ResourceConfig, ConnectorState]):
        self._buffer.write(
            response.model_dump_json(by_alias=True, exclude_unset=True).encode()
        )
        self._buffer.write(b"\n")
        self._buffer.seek(0)
        shutil.copyfileobj(self._buffer, self._output)
        self._output.flush()
        self.reset()


class BaseCaptureConnector(
    BaseConnector[Request[EndpointConfig, ResourceConfig, ConnectorState]],
    Generic[EndpointConfig, ResourceConfig, ConnectorState],
):
    output: BinaryIO = sys.stdout.buffer

    @abc.abstractmethod
    async def spec(self, log: Logger, _: request.Spec) -> ConnectorSpec:
        raise NotImplementedError()

    @abc.abstractmethod
    async def discover(
        self,
        log: Logger,
        discover: request.Discover[EndpointConfig],
    ) -> response.Discovered[ResourceConfig]:
        raise NotImplementedError()

    @abc.abstractmethod
    async def validate(
        self,
        log: Logger,
        validate: request.Validate[EndpointConfig, ResourceConfig],
    ) -> response.Validated:
        raise NotImplementedError()

    async def apply(
        self,
        log: Logger,
        apply: request.Apply[EndpointConfig, ResourceConfig],
    ) -> response.Applied:
        return response.Applied(actionDescription="")

    @abc.abstractmethod
    async def open(
        self,
        log: Logger,
        open: request.Open[EndpointConfig, ResourceConfig, ConnectorState],
    ) -> tuple[response.Opened, Callable[[Task], Awaitable[None]]]:
        raise NotImplementedError()

    async def acknowledge(self, acknowledge: request.Acknowledge) -> None:
        return None  # No-op.

    async def handle(
        self,
        log: Logger,
        request: Request[EndpointConfig, ResourceConfig, ConnectorState],
    ) -> None:

        if spec := request.spec:
            response = await self.spec(log, spec)
            response.protocol = 3032023
            self._emit(Response(spec=response))

        elif discover := request.discover:
            self._emit(Response(discovered=await self.discover(log, discover)))

        elif validate := request.validate_:
            self._emit(Response(validated=await self.validate(log, validate)))

        elif apply := request.apply:
            self._emit(Response(applied=await self.apply(log, apply)))

        elif open := request.open:
            opened, capture = await self.open(log, open)
            self._emit(Response(opened=opened))

            stopping = Task.Stopping(asyncio.Event())

            async def stop_on_elapsed_interval(interval: int) -> None:
                await asyncio.sleep(interval)
                stopping.event.set()

            # Gracefully exit after the capture interval has elapsed.
            # We don't do this within the TaskGroup because we don't
            # want to block on it.
            asyncio.create_task(stop_on_elapsed_interval(open.capture.intervalSeconds))

            async with asyncio.TaskGroup() as tg:

                task = Task(
                    log.getChild("capture"),
                    "capture",
                    self.output,
                    stopping,
                    tg,
                )
                await capture(task)

            # When capture() completes, the connector exits.
            if stopping.first_error:
                raise Stopped(
                    f"Task {stopping.first_error_task}: {stopping.first_error}"
                )
            else:
                raise Stopped(None)

        elif acknowledge := request.acknowledge:
            await self.acknowledge(acknowledge)

        else:
            raise RuntimeError("malformed request", request)

    def _emit(self, response: Response[EndpointConfig, ResourceConfig, ConnectorState]):
        self.output.write(
            response.model_dump_json(by_alias=True, exclude_unset=True).encode()
        )
        self.output.write(b"\n")
        self.output.flush()
