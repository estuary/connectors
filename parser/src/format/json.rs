//! Parser for the json format. This will accept any stream of JSON values separated by whitespace.
//! It allows any amount of whitespace (including newlines) within and in between records.
use super::{Input, Output, ParseError, Parser};
use crate::ParseConfig;

struct JsonParser;

pub fn new_parser() -> Box<dyn Parser> {
    Box::new(JsonParser)
}

impl Parser for JsonParser {
    fn parse(&self, _config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
        let deser = serde_json::de::Deserializer::from_reader(content).into_iter();
        let wrapped = deser.map(|r| r.map_err(Into::into));
        Ok(Box::new(wrapped))
    }
}
