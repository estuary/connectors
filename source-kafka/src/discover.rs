use std::collections::HashMap;

use anyhow::{Context, Result};
use apache_avro::schema::Schema as AvroSchema;
use doc::{
    shape::{schema::to_schema, ObjProperty},
    Shape,
};
use json::schema::{self as JsonSchema, types};
use proto_flow::capture::{request::Discover, response::discovered};
use rdkafka::consumer::Consumer;
use schemars::schema::RootSchema;
use serde_json::json;

use crate::{
    configuration::{EndpointConfig, Resource, SchemaRegistryConfig},
    schema_registry::{
        RegisteredSchema::{Avro, Json, Protobuf},
        SchemaRegistryClient, TopicSchema,
    },
    KAFKA_TIMEOUT,
};

static KAFKA_INTERNAL_TOPICS: [&str; 3] = ["__consumer_offsets", "__amazon_msk_canary", "_schemas"];

pub async fn do_discover(req: Discover) -> Result<Vec<discovered::Binding>> {
    let config: EndpointConfig = serde_json::from_str(&req.config_json)?;
    let consumer = config.to_consumer().await?;

    let meta = consumer
        .fetch_metadata(None, KAFKA_TIMEOUT)
        .context("Could not connect to bootstrap server with the provided configuration. This may be due to an incorrect configuration for authentication or bootstrap servers. Double check your configuration and try again.")?;

    let mut all_topics: Vec<String> = meta
        .topics()
        .iter()
        .filter_map(|t| {
            let name = t.name();
            if KAFKA_INTERNAL_TOPICS.contains(&name) {
                None
            } else {
                Some(name.to_string())
            }
        })
        .collect();

    all_topics.sort();

    let registered_schemas = match config.schema_registry {
        SchemaRegistryConfig::ConfluentSchemaRegistry {
            endpoint,
            username,
            password,
        } => {
            let client = SchemaRegistryClient::new(endpoint, username, password);
            client
                .schemas_for_topics(&all_topics)
                .await
                .context("Could not connect to the configured schema registry. Double check your configuration and try again.")?
        }
        SchemaRegistryConfig::NoSchemaRegistry { .. } => HashMap::new(),
    };

    all_topics
        .into_iter()
        .filter_map(|topic| {
            let registered_schema = match registered_schemas.get(&topic) {
                Some(s) => s,
                None => &TopicSchema::default(),
            };

            if matches!(&registered_schema.key, Some(Protobuf))
                || matches!(&registered_schema.value, Some(Protobuf))
            {
                // TODO(whb): At some point we may want to support protobuf
                // schemas.
                return None;
            }

            let (collection_schema, key_ptrs) =
                match topic_schema_to_collection_spec(registered_schema) {
                    Ok(s) => s,
                    Err(e) => return Some(Err(e)),
                };

            Some(Ok(discovered::Binding {
                recommended_name: topic.to_owned(),
                resource_config_json: serde_json::to_string(&Resource {
                    topic: topic.to_owned(),
                })
                .expect("resource config must serialize"),
                document_schema_json: serde_json::to_string(&collection_schema)
                    .expect("document schema must serialize"),
                key: key_ptrs,
                resource_path: vec![topic.to_owned()],
                ..Default::default()
            }))
        })
        .collect::<Result<Vec<_>>>()
}

fn topic_schema_to_collection_spec(
    topic_schema: &TopicSchema,
) -> Result<(RootSchema, Vec<String>)> {
    let mut collection_key = vec!["/_meta/partition".to_string(), "/_meta/offset".to_string()];
    let doc_schema_json = json!({
        "x-infer-schema": true,
        "if": {
            "properties": {
              "_meta": {
                "properties": {
                  "op": {
                    "const": "d"
                  }
                }
              }
            }
          },
        "then": {
            "reduce": {
                "delete": true,
                "strategy": "merge"
            }
        },
        "type": "object",
        "properties": {
            "_meta": {
                "type": "object",
                "properties": {
                    "topic": {
                        "description": "The topic the message was read from",
                        "type": "string",
                    },
                    "partition": {
                        "description": "The partition the message was read from",
                        "type": "integer",
                    },
                    "offset": {
                        "description": "The offset of the message within the partition",
                        "type": "integer",
                    },
                    "op": {
                        "enum": ["c", "u", "d"],
                        "description": "Change operation type: 'c' Create/Insert, 'u' Update, 'd' Delete.",
                    }
                },
                "required": ["offset", "op", "partition", "topic"]
            }
        },
        "required": ["_meta"]
    });

    let mut collection_schema: RootSchema = serde_json::from_value(doc_schema_json).unwrap();

    let mut key_shape = match &topic_schema.key {
        Some(Avro(schema)) => avro_key_schema_to_shape(schema)?,
        Some(Json(schema)) => json_key_schema_to_shape(schema)?,
        Some(Protobuf) => todo!("protobuf schemas are not yet supported"),
        None => Shape::nothing(),
    };

    if usable_key_shape(&key_shape) {
        if key_shape.type_ != JsonSchema::types::OBJECT {
            // The topic key is a single value not part of a record, and
            // that cannot be represented as a JSON document. There has
            // to be a key/value pair in the document, so transmute this
            // unnamed scalar value into a document with a synthetic key
            // to reference its value.
            let scalar_shape = key_shape.clone();
            key_shape = Shape::nothing();
            key_shape.type_ = JsonSchema::types::OBJECT;
            key_shape.object.properties = vec![ObjProperty {
                name: "_key".into(),
                is_required: true,
                shape: scalar_shape,
            }];
        }

        collection_key = key_shape
            .locations()
            .into_iter()
            .filter(|(ptr, _, shape, ..)| {
                !shape.type_.overlaps(JsonSchema::types::OBJECT) && ptr.to_string() != "/*"
            })
            .map(|(ptr, ..)| ptr.to_string())
            .collect();

        let mut key_schema = to_schema(key_shape);

        collection_schema
            .schema
            .object()
            .properties
            .append(&mut key_schema.schema.object().properties);

        collection_schema
            .schema
            .object()
            .required
            .append(&mut key_schema.schema.object().required);
    }

    Ok((collection_schema, collection_key))
}

fn avro_key_schema_to_shape(schema: &AvroSchema) -> Result<Shape> {
    let mut shape = Shape::nothing();

    shape.type_ = match schema {
        AvroSchema::Boolean => JsonSchema::types::BOOLEAN,
        AvroSchema::Int | AvroSchema::Long => JsonSchema::types::INTEGER,
        AvroSchema::String => JsonSchema::types::STRING,
        AvroSchema::Bytes | AvroSchema::Fixed(_) => {
            shape.string.content_encoding = Some("base64".into());
            JsonSchema::types::STRING
        }
        AvroSchema::Decimal(_) | AvroSchema::BigDecimal => {
            shape.string.format = Some(JsonSchema::formats::Format::Number);
            JsonSchema::types::STRING
        }
        AvroSchema::Uuid => {
            shape.string.format = Some(JsonSchema::formats::Format::Uuid);
            JsonSchema::types::STRING
        }
        AvroSchema::Date => {
            shape.string.format = Some(JsonSchema::formats::Format::Date);
            JsonSchema::types::STRING
        }
        AvroSchema::TimeMillis | AvroSchema::TimeMicros => {
            shape.string.format = Some(JsonSchema::formats::Format::Time);
            JsonSchema::types::STRING
        }
        AvroSchema::TimestampMillis
        | AvroSchema::TimestampMicros
        | AvroSchema::TimestampNanos
        | AvroSchema::LocalTimestampMillis
        | AvroSchema::LocalTimestampMicros
        | AvroSchema::LocalTimestampNanos => {
            shape.string.format = Some(JsonSchema::formats::Format::DateTime);
            JsonSchema::types::STRING
        }
        AvroSchema::Duration => {
            shape.string.format = Some(JsonSchema::formats::Format::Duration);
            JsonSchema::types::STRING
        }
        AvroSchema::Enum(enum_schema) => {
            shape.enum_ = Some(
                enum_schema
                    .symbols
                    .iter()
                    .map(|s| s.to_string().into())
                    .collect(),
            );
            JsonSchema::types::STRING
        }
        AvroSchema::Record(record_schema) => {
            shape.object.properties = record_schema
                .fields
                .iter()
                .map(|field| {
                    let mut field_shape = avro_key_schema_to_shape(&field.schema)?;

                    if let Some(doc) = &field.doc {
                        field_shape.description = Some(doc.to_string().into());
                    }
                    if let Some(default) = &field.default {
                        field_shape.default = Some((default.to_owned(), None).into());
                    }

                    Ok(ObjProperty {
                        name: field.name.clone().into(),
                        is_required: field.default.is_none(),
                        shape: field_shape,
                    })
                })
                .collect::<Result<Vec<_>>>()?;
            JsonSchema::types::OBJECT
        }

        // Schemas that allow 'null' are not schematized as keys since
        // nullable keys are not very useful in practice.
        AvroSchema::Null => JsonSchema::types::INVALID,

        // Similarly, schemas with multiple types or a single type with an
        // additional explicit 'null' are not very useful in practice,
        // although technically allowed by Flow, so they won't be
        // schematized.
        AvroSchema::Union(_) => JsonSchema::types::INVALID,

        // We could perhaps treat floating points as string-encoded numbers,
        // but if that's what they were then they should probably specified
        // as decimals. And this is more consistent with what is achievable
        // with JSON schemas.
        AvroSchema::Float => JsonSchema::types::INVALID,
        AvroSchema::Double => JsonSchema::types::INVALID,

        // Arrays and maps just don't make any sense as key schemas.
        AvroSchema::Array(_) => JsonSchema::types::INVALID,
        AvroSchema::Map(_) => JsonSchema::types::INVALID,

        AvroSchema::Ref { name } => anyhow::bail!("Avro key schema contains reference {}", name),
    };

    Ok(shape)
}

fn json_key_schema_to_shape(schema: &serde_json::Value) -> Result<Shape> {
    let json_schema = doc::validation::build_bundle(&schema.to_string())?;
    let validator = doc::Validator::new(json_schema)?;
    Ok(doc::Shape::infer(
        &validator.schemas()[0],
        validator.schema_index(),
    ))
}

fn usable_key_shape(shape: &Shape) -> bool {
    // Schemas may be valid keys if all the properties are keyable and
    // non-nullable, including nested properties. Non-nullable here means they
    // can't be an explicit null, and either must exist or have a default value.
    shape
        .locations()
        .iter()
        .all(|(ptr, pattern, shape, exists)| {
            if ptr.to_string() == "/*" && shape.type_ == types::INVALID {
                // This represents an "additionalProperties: false"
                // configuration, which must be the case for a valid key schema.
                return true;
            }

            !pattern
                && (exists.must() || shape.default.is_some())
                && !shape.type_.overlaps(JsonSchema::types::NULL)
                && (shape.type_.is_keyable_type() || shape.type_ == types::OBJECT)
        })
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_topic_schema_to_collection_spec() {
        let test_cases = [
            (
                "no key",
                &TopicSchema {
                    key: None,
                    ..Default::default()
                },
            ),
            (
                "single scalar avro key",
                &TopicSchema {
                    key: Some(Avro(
                        apache_avro::Schema::parse(&json!({"type": "string"})).unwrap(),
                    )),
                    ..Default::default()
                },
            ),
            (
                "single nullable scalar avro key",
                &TopicSchema {
                    key: Some(Avro(
                        apache_avro::Schema::parse(&json!({"type": ["null", "string"]})).unwrap(),
                    )),
                    ..Default::default()
                },
            ),
            (
                "single non-scalar avro key",
                &TopicSchema {
                    key: Some(Avro(
                        apache_avro::Schema::parse(&json!({"type": "array", "items": "string"}))
                            .unwrap(),
                    )),
                    ..Default::default()
                },
            ),
            (
                "avro record with scalar compound key",
                &TopicSchema {
                    key: Some(Avro(
                        apache_avro::Schema::parse(&json!({
                            "type": "record",
                            "name": "someRecord",
                            "fields": [
                                {"name": "firstKey", "type": "string", "doc": "the first key field"},
                                {"name": "secondKey", "type": "long", "doc": "the second key field"},
                                {
                                    "name": "thirdKey",
                                    "type": "enum",
                                    "symbols": ["a", "b", "c"],
                                    "default": "a",
                                    "doc": "the third key field, which is an enum with a default value",
                                },
                            ],
                        }))
                        .unwrap(),
                    )),
                    ..Default::default()
                },
            ),
            (
                "nested avro record with scalars",
                &TopicSchema {
                    key: Some(Avro(
                        apache_avro::Schema::parse(&json!({
                            "type": "record",
                            "name": "someRecord",
                            "fields": [
                                {"name": "firstKey", "type": "string", "doc": "the first key field"},
                                {
                                    "name": "nestedRecord",
                                    "type": "record",
                                    "fields": [
                                        {"name": "secondKeyNested", "type": "long", "doc": "the second key field"},
                                        {"name": "thirdKeyNested", "type": "bytes", "doc": "the third key field"},
                                    ],
                                },
                            ],
                        }))
                        .unwrap(),
                    )),
                    ..Default::default()
                },
            ),
            (
                "single scalar json key",
                &TopicSchema {
                    key: Some(Json(json!({"type": "string"}))),
                    ..Default::default()
                },
            ),
            (
                "single nullable scalar json key",
                &TopicSchema {
                    key: Some(Json(json!({"type": ["null", "string"]}))),
                    ..Default::default()
                },
            ),
            (
                "nested json object",
                &TopicSchema {
                    key: Some(Json(json!({
                        "type": "object",
                        "properties": {
                            "firstKey": {"type": "string"},
                            "nestedObject": {
                                "type": "object",
                                "properties": {
                                    "secondKeyNested": {"type": "integer"},
                                    "thirdKeyNested": {"type": "boolean"},
                                },
                                "required": ["secondKeyNested", "thirdKeyNested"],
                            },
                        },
                        "required": ["firstKey", "nestedObject"],
                        "additionalProperties": false
                    }))),
                    ..Default::default()
                },
            ),
        ];

        for (name, input) in test_cases {
            let (discovered_schema, discovered_key) =
                topic_schema_to_collection_spec(input).unwrap();
            let mut snap = String::new();
            snap.push_str(&serde_json::to_string(&discovered_key).unwrap());
            snap.push_str("\n");
            snap.push_str(&serde_json::to_string_pretty(&discovered_schema).unwrap());
            assert_snapshot!(name, snap);
        }
    }

    #[test]
    fn test_avro_key_schema_to_shape() {
        let test_cases = [
            (
                vec![json!({"type": "boolean"})],
                Some(json!({"type": "boolean"})),
            ),
            (
                vec![json!({"type": "int"}), json!({"type": "long"})],
                Some(json!({"type": "integer"})),
            ),
            (
                vec![json!({"type": "string"})],
                Some(json!({"type": "string"})),
            ),
            (
                vec![
                    json!({"type": "bytes"}),
                    json!({"type": "fixed", "name": "foo", "size": 10}),
                ],
                Some(json!({"type": "string", "contentEncoding": "base64"})),
            ),
            (
                vec![
                    json!({"type": "bytes", "logicalType": "decimal", "precision": 4, "scale": 2}),
                    json!({"type": "bytes", "logicalType": "big-decimal", "precision": 4, "scale": 2}),
                ],
                Some(json!({"type": "string", "format": "number"})),
            ),
            (
                vec![json!({"type": "string", "logicalType": "uuid"})],
                Some(json!({"type": "string", "format": "uuid"})),
            ),
            (
                vec![json!({"type": "int", "logicalType": "date"})],
                Some(json!({"type": "string", "format": "date"})),
            ),
            (
                vec![
                    json!({"type": "int", "logicalType": "time-millis"}),
                    json!({"type": "long", "logicalType": "time-micros"}),
                ],
                Some(json!({"type": "string", "format": "time"})),
            ),
            (
                vec![
                    json!({"type": "long", "logicalType": "timestamp-millis"}),
                    json!({"type": "long", "logicalType": "timestamp-micros"}),
                    json!({"type": "long", "logicalType": "local-timestamp-millis"}),
                    json!({"type": "long", "logicalType": "local-timestamp-micros"}),
                    // TODO(whb): These nanosecond timestamps are not parsed by
                    // the Avro library and I'm not sure if they are really part
                    // of the spec.
                    // json!({"type": "long", "logicalType": "timestamp-nanos"})),
                    // json!({"type": "long", "logicalType": "local-timestamp-nanos"})),
                ],
                Some(json!({"type": "string", "format": "date-time"})),
            ),
            (
                vec![
                    json!({"type": "fixed", "name": "foo", "size": 12, "logicalType": "duration"}),
                ],
                Some(json!({"type": "string", "format": "duration"})),
            ),
            (
                vec![json!({"type": "enum", "name": "foo", "symbols": ["a", "b", "c"]})],
                Some(json!({"type": "string", "enum": ["a", "b", "c"]})),
            ),
            (
                vec![
                    json!({"type": "null"}),
                    json!({"type": ["null", "long"]}),
                    json!({"type": "float"}),
                    json!({"type": "double"}),
                    json!({"type": "array", "items": "string"}),
                    json!({"type": "map", "values": "string"}),
                ],
                None,
            ),
        ];

        for (schema_jsons, want) in test_cases {
            for schema_json in schema_jsons {
                let shape =
                    avro_key_schema_to_shape(&AvroSchema::parse(&schema_json).unwrap()).unwrap();
                if usable_key_shape(&shape) {
                    assert_eq!(
                        serde_json::to_value(&to_schema(shape).schema).unwrap(),
                        serde_json::to_value(&want.clone().unwrap()).unwrap()
                    )
                } else {
                    assert!(want.is_none())
                }
            }
        }
    }
}
