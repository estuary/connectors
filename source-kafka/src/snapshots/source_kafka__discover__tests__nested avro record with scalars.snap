---
source: src/discover.rs
expression: snap
---
["/firstKey","/nestedRecord/secondKeyNested","/nestedRecord/thirdKeyNested"]
{
  "type": "object",
  "required": [
    "_meta",
    "firstKey",
    "nestedRecord"
  ],
  "properties": {
    "_meta": {
      "type": "object",
      "required": [
        "offset",
        "partition"
      ],
      "properties": {
        "offset": {
          "description": "The offset of the message within the partition",
          "type": "integer"
        },
        "partition": {
          "description": "The partition the message was read from",
          "type": "integer"
        },
        "topic": {
          "description": "The topic the message was read from",
          "type": "string"
        }
      }
    },
    "firstKey": {
      "description": "the first key field",
      "type": "string"
    },
    "nestedRecord": {
      "type": "object",
      "required": [
        "secondKeyNested",
        "thirdKeyNested"
      ],
      "properties": {
        "secondKeyNested": {
          "description": "the second key field",
          "type": "integer"
        },
        "thirdKeyNested": {
          "description": "the third key field",
          "type": "string",
          "contentEncoding": "base64"
        }
      }
    }
  },
  "x-infer-schema": true
}
